### PR TITLE
CSHARP-3102: Memory leak related to heartbeat when connecting to Atlas clusters

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
https://jira.mongodb.org/browse/CSHARP-3102
https://evergreen.mongodb.com/version/5f1070762fbabe26ddfd55a3

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs [hunk 1]
```diff
@@ -222,7 +222,16 @@ private IBsonSerializer CreateSizeLimitingPayloadSerializer(Type1CommandMessageS
...
+            try
+            {
+                connection.ReceiveMessageAsync(message.RequestId, encoderSelector, _messageEncoderSettings, cancellationToken)
+                    .GetAwaiter()
+                    .GetResult();
+            }
+            catch
+            {
+                // ignored
+            }
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs]: `IgnoreResponse` is a sync method and should be 100% sync without calling any async methods:

    private void IgnoreResponse(IConnection connection, QueryMessage message, CancellationToken cancellationToken)
    {
        var encoderSelector = new ReplyMessageEncoderSelector<IgnoredReply>(IgnoredReplySerializer.Instance);
        try
        {
            connection.ReceiveMessage(message.RequestId, encoderSelector, _messageEncoderSettings, cancellationToken);
        }
        catch
        {
            // ignore exceptions
        }
    }

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs]: And you should add a 100% async version that can be called from `ExecuteAsync`:

    private async Task IgnoreResponseAsync(IConnection connection, QueryMessage message, CancellationToken cancellationToken)
    {
        var encoderSelector = new ReplyMessageEncoderSelector<IgnoredReply>(IgnoredReplySerializer.Instance);
        try
        {
            await connection.ReceiveMessageAsync(message.RequestId, encoderSelector, _messageEncoderSettings, cancellationToken).ConfigureAwait(false);
        }
        catch
        {
            // ignore exceptions
        }
    }

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs]: Line 148 in `ExecuteAsync` should be:

    await IgnoreResponseAsync(connection, message, cancellationToken).ConfigureAwait(false);

**Review by rstam — CHANGES_REQUESTED** [review]
Small changes requested.

As I think I heard you mention already that these changes make unacknowledged writes a BLOCKING operation when using OP_QUERY (i.e. servers older than 3.6).

I don't consider that to be a problem though... @jyemin?

File: src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs (hunk 1 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs]: Great suggestion! I didn't realize there's an async version. Now the code looks much better.
Done.

**Review by MikalaiMazurenka — COMMENTED** [review]
I've addressed all the suggestions. CommandWriteProtocolTests.cs required an update as well.
Ready for review!

**Comment by jyemin** [comment]
> As I think I heard you mention already that these changes make unacknowledged writes a BLOCKING operation when using OP_QUERY (i.e. servers older than 3.6).

@MikalaiMazurenka can you clarify what is meant by "blocking" here?  What is the precise behavioral change?  I thought it was that the connection won't be returned to the pool until the write operation completes.  Does it also block the application thread until the write operation completes?

**Comment by MikalaiMazurenka** [comment]
> @MikalaiMazurenka can you clarify what is meant by "blocking" here? What is the precise behavioral change? I thought it was that the connection won't be returned to the pool until the write operation completes. Does it also block the application thread until the write operation completes?

For the async version (`ExecuteAsync`) "blocking" in this context means execution flow block. It doesn't block the thread, because when we hit await, the thread which was executing this would be returned to thread pool.  What I mean by "execution flow block" is that 
```
return default(TCommandResult);
```
Line wouldn't be hit until
```
await IgnoreResponseAsync(connection, message, cancellationToken).ConfigureAwait(false);
```
Line completes (which means receiving the message). Still, it wouldn't block any threads because the latter is an asynchronous call.

For the synchronous version this would block execution flow and current thread.

For both sync and async versions this call was "fire and forget" which executed as a separate task and the execution of `ExecuteAsync` or `Execute` continued as soon as we call the `IgnoreResponse` (without waiting for the receive to complete). With my changes those methods would block the method execution until `IgnoreResponseAsync` or `IgnoreResponse` completes and would block the thread for synchronous flow.

**Comment by jyemin** [comment]
> I don't consider that to be a problem though... 

It means than on 2.6 - 3.4 servers, the driver has silently disabled unacknowledged writes.  It's also different from all other drivers that I'm aware of, which handle this by using the old wire protocol (OP_INSERT/OP_UPDATE/OP_DELETE to support unacknowledged writes on 2.6-3.4 servers.  Given that, I do consider it to be a problem, and a reason for not proceeding with this change.

**Comment by rstam** [comment]
> It means than on 2.6 - 3.4 servers, the driver has silently disabled unacknowledged writes. It's also different from all other drivers that I'm aware of, which handle this by using the old wire protocol (OP_

Or... it means that we need  to proceed with this change with some additional work. For example, a connection could maintain a list or response IDs that it should ignore (when they eventually come in). Then the connection can be safely put back in the connection pool and everything should work pretty much the same way as it does now.

Not proceeding with this change means we haven't fixed the memory leak, and fixing the memory leak in the Dropbox related code might be rather difficult.

**Comment by jyemin** [comment]
That's a different change.  I meant that we should not proceed with the change as is.

**Comment by MikalaiMazurenka** [comment]
I've reverted the `DropBox` removal and updated only the code responsible for memory leak.

**Comment by rstam** [comment]
This change might fix the memory leak but it now blocks on unacknowledged writes, which means that we are no longer supporting the key feature of unacknowledged writes which is not having to wait for a reply.

**Pull request closed without merging**